apps/CA.pl.in

```diff
@@ -127,59 +162,289 @@ if ($WHAT eq '-newcert' ) {
+}
+
+# Update Certificate Revocation List
+sub gencrl
+{
+    $RET = run("$CA $OPENSSL_CONFIG -gencrl -out ${CATOP}/crl/$CACRL");
+    if ($RET != 0) {
+        print STDERR "Failed to Generated CRL\n";
+        exit 1;
+    }
+    print "Generated CRL is in ${CATOP}/crl/$CACRL\n";
+    # Place what to do after CRL update, e.g.
+    # $RET = run("cd ${CERTTOP} && make");
+}
+
+# Revoke a certificate
+sub revoke
+{
+    # revoke a certificate
+    my ($cname,$reason) = @_;
+    unless ($cname) {
         print "Certificate filename is required; reason optional.\n";
         exit 1;
     }
-    my $reason = $ARGV[2];
-    $reason = " -crl_reason $reason"
-        if defined $reason && crl_reason_ok($reason);
-    $RET = run("$CA -revoke \"$cname\"" . $reason);
+    $reason = "" unless defined $reason;
+    if (length($reason) gt 0 && crl_reason_ok($reason)) {
+        $reason = " -crl_reason $reason"
+    }
+    unless (-f "$cname") {
+        print STDERR "No such file $cname\n";
+        exit 1;
+    }
+    $RET = run("$CA $OPENSSL_CONFIG -revoke \"$cname\"" . $reason);
+    if ($RET != 0) {
+        print STDERR "Failed to revoke $cname\n";
+        exit 1;
+    }
+}
+
+# Revoke and delete a set of cert, key and PKCS#12 files
+sub p12mng_revoke
+{
+    $OPENSSL_CONFIG = $OPENSSL_CONFIG_P12MNG unless $OPENSSL_CONFIG;
+    my ($cname,$reason) = @_;
+    unless ($cname) {
+        print "Certificate name is required; reason optional.\n";
+        exit 1;
+    }
+    $reason = "" unless defined $reason;
+    unless (-f "${CERTTOP}/certs/${cname}.pem") {
+        print STDERR "No such file ${cname}.pem\n";
+        exit 1;
+    }
+    revoke("${CERTTOP}/certs/${cname}.pem",$reason);
+    # remove the revoked key set
+    run("rm -i ${CERTTOP}/certs/${cname}.pem");
+    run("rm -i ${CERTTOP}/private/${cname}.key");
+    run("rm -i ${CERTTOP}/private/${cname}.p12");
+    print "cname $cname successfully revoked.\n";
+    gencrl();
+}
+
+
+unless ($WHAT) {
+    help();
+    exit 0;
+}
+if ( $WHAT =~ /^(-\?|-h|-help)$/ ) {
+    help();
+    exit 0;
+}
+if ($WHAT eq '-newcert' ) {
+    newcert();
+} elsif ($WHAT eq '-newreq' ) {
+    newreq();
+} elsif ($WHAT eq '-newreq-nodes' ) {
+    newreq_nodes();
+} elsif ($WHAT eq '-newca' ) {
+    newca();
+} elsif ($WHAT eq '-pkcs12' ) {
+    pkcs12($ARGV[0]);
+} elsif ($WHAT eq '-xsign' ) {
+    xsign($ARGV[0],$ARGV[1]);
+} elsif ($WHAT eq '-sign' ) {
+    sign($ARGV[0],$ARGV[1]);
+} elsif ($WHAT eq '-signCA' ) {
+    signCA();
+} elsif ($WHAT eq '-signcert' ) {
+    signcert($ARGV[0],$ARGV[1]);
+} elsif ($WHAT eq '-selfsign' ) {
+    selfsign($ARGV[0],$ARGV[1]);
+} elsif ($WHAT eq '-verify' ) {
+    verify();
+} elsif ($WHAT eq '-p12mng-new' ) {
+    p12mng_new($ARGV[0],$ARGV[1]);
+} elsif ($WHAT eq '-crl' ) {
+    gencrl();
+} elsif ($WHAT eq '-revoke' ) {
+    revoke($ARGV[0],$ARGV[1]);
+} elsif ($WHAT eq '-p12mng-revoke' ) {
+    p12mng_revoke($ARGV[0],$ARGV[1]);
```

Comment:
In all these cases, I think it would be better to pass `@ARGV` instead of a selection of it, and properly parse the arguments in each function.